r/src/compute.cpp

```diff
@@ -82,4 +82,148 @@ std::shared_ptr<arrow::Table> Table__cast(
   return arrow::Table::Make(schema, std::move(columns), table->num_rows());
 }
 
+// [[arrow::export]]
+std::shared_ptr<arrow::Array> Array__Take(
+    const std::shared_ptr<arrow::Array>& values,
+    const std::shared_ptr<arrow::Array>& indices) {
+
+  std::shared_ptr<arrow::Array> out;
+  arrow::compute::FunctionContext context;
+  arrow::compute::TakeOptions options;
+  STOP_IF_NOT_OK(arrow::compute::Take(&context, *values, *indices, options, &out));
+  return out;
+}
+
+// [[arrow::export]]
+std::shared_ptr<arrow::RecordBatch> RecordBatch__Take(
+    const std::shared_ptr<arrow::RecordBatch>& batch,
+    const std::shared_ptr<arrow::Array>& indices) {
+  int ncols = batch->num_columns();
+  auto nrows = indices->length();
+
+  std::vector<std::shared_ptr<arrow::Array>> columns(ncols);
+
+  for (R_xlen_t j = 0; j < ncols; j++) {
+    columns[j] = Array__Take(batch->column(j), indices);
+  }
+
+  return arrow::RecordBatch::Make(batch->schema(), nrows, columns);
+}
+
+// [[arrow::export]]
+std::shared_ptr<arrow::ChunkedArray> ChunkedArray__Take(
+    const std::shared_ptr<arrow::ChunkedArray>& values,
+    Rcpp::IntegerVector& indices) {
+
+  int num_chunks = values->num_chunks();
+  std::vector<std::shared_ptr<arrow::Array>> new_chunks(1); // Hard-coded 1 for now
+  // TODO: fast path if there's only 1 chunk?
+
+  std::shared_ptr<arrow::Array> current_chunk;
+  std::shared_ptr<arrow::Array> current_indices;
+  int offset = 0;
+  int len;
+  int min_i = indices[0];
+  int max_i = indices[0];
+
+  // 1) see if all i are in the same chunk, call Array__Take on that
+  for (R_xlen_t i = 0; i < indices.size(); i++) {
+    if (indices[i] < min_i) {
+      min_i = indices[i];
+    } else if (indices[i] > max_i) {
+      max_i = indices[i];
+    }
+  }
+
+  for (R_xlen_t chk = 0; chk < num_chunks; chk++) {
+    current_chunk = values->chunk(chk);
+    len = current_chunk->length();
+    if (min_i >= offset & max_i < offset + len) {
+      for (R_xlen_t i = 0; i < indices.size(); i++) {
+        // Subtract offset from all indices
+        indices[i] -= offset;
+      }
+      current_indices = arrow::r::Array__from_vector(indices, arrow::int32(), true);
+      new_chunks[0] = Array__Take(current_chunk, current_indices);
+      return std::make_shared<arrow::ChunkedArray>(std::move(new_chunks));
+    }
+    offset += len;
+  }
+  Rcpp::stop("ChunkedArray$Take() only supported when all i are in the same chunk");
```

Comment:
Alternatively, you could concatenate the chunks and call Take on the result:
```suggestion
  std::shared_ptr<Array> concatenated;
  RETURN_NOT_OK(arrow::Concatenate(values->chunks(), arrow::default_memory_pool(), &concatenated));
  new_chunks[0] = Array__Take(concatenated, arrow::r::Array__from_vector(indices, arrow::int32(), true));
  return std::make_shared<arrow::ChunkedArray>(std::move(new_chunks));
```

Reply:
Is concatenating expensive or even always possible? I'd expect that one reason you'd have a ChunkedArray is that for some reason the values won't all fit into a single Array.

I'm fine with doing "slow and naive but works" to start and optimizing later, but I'm just not clear on what ChunkedArrays are about and what we're allowed to do with them.

Reply:
Concatenating is relatively expensive; it uses as much memory as all the chunks in your array again. If you don't have enough memory it'll return an error status (instead of crashing) but apart from that it's always possible.

Concatenating chunks will never change the value in a slot; it's just expensive so we try to avoid it. I would definitely say you're allowed to concatenate wherever you are working on the initial implementation for chunked arrays.

Reply:
Cool. I'll try that. I'd rather ship something that doesn't have seemingly arbitrary restrictions on how to use it, and we can explore performance etc. in https://issues.apache.org/jira/browse/ARROW-6784

Reply:
I did this here: https://github.com/apache/arrow/pull/5531/commits/038e2a846279c2258229590f1c9c78a3da024ee2#diff-ffc5e6f7dfee7f9bed7733b7fb41e632R161-R166